File: web2-project/src/app/model/work-plan.ts
import { Address } from "./address";
import { Device } from "./device";
import { Incident } from "./incident";
import { Team } from "./team/team.model";
import { User } from "./user";
import { WorkAccount } from "./work-account";
import { WorkInstruction } from "./work-instruction";

export enum Status{
    Draft,
    Approved,
    Denied,
    Canceled,
    Finished
}

export enum Type{
    Planned,
    Unplanned
}

export class WorkPlan {// treba da se doda nalog za rad
    public workPlanID: number;
    public type: Type;
    public workAccount?: WorkAccount;
    public status: Status;
    public incident?: Incident;
    public address: Address;
    public team: Team;
    public createdBy: string;
    public startDate: Date;
    public endDate: Date;
    public purpose: string;
    public notes: string;
    public company: string;
    public phone: number;
    public creationDate: Date;
    public statusHistory: Array<WorkPlanStatusHistory>;
    public images: Array<string>;
    public equipment : number;
    public instructions: Array<WorkInstruction>;

    constructor(type: Type, address: Address, createdBy: string, startDate: Date, endDate: Date, purpose: string, notes: string, company: string, phone: number, creationDate: Date, incident?: Incident, workAccount?: WorkAccount){
        this.type = type;
        this.workAccount = workAccount;
        this.status = Status.Draft;
        this.incident = incident;
        this.address = address;
        //this.team = team;
        this.createdBy = createdBy;
        this.startDate = startDate;
        this.endDate = endDate;
        this.purpose = purpose;
        this.notes = notes;
        this.company = company;
        this.phone = phone;
        this.creationDate = creationDate;
        this.statusHistory = new Array<WorkPlanStatusHistory>();
        this.statusHistory.push(new WorkPlanStatusHistory(this.creationDate, this.createdBy, this.status, this.workPlanID));
        this.instructions = new Array<WorkInstruction>();
    }
}

export class WorkPlanStatusHistory{
    public workPlanStatusHistoryID: number;
    public date: Date;
    public changedBy: string;
    public status: Status;
    public workPlanID: number;

    constructor(date: Date, user: string, status: Status, workPlanID: number){
        this.date = date;
        this.changedBy = user;
        this.status = status;
        this.workPlanID = workPlanID;
    }
}

File: Web2BackEnd/Web2BackEnd/Models/WorkPlan.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace Web2BackEnd.Models
{
	public class WorkPlan : GenericModel
	{
		[Key, DatabaseGenerated(DatabaseGeneratedOption.Identity)]
		public int WorkPlanID { get; set; }
		[Required]
		public WorkPlanType Type { get; set; }
		[Required]
		public WorkPlanStatus Status { get; set; }
		[Required]
		public Address Address { get; set; }
		[Required]
		public string CreatedBy { get; set; }
		[Required]
		public DateTime StartDate { get; set; }
		[Required]
		public DateTime EndDate { get; set; }
		[Required]
		public string Purpose { get; set; }
		public string Notes { get; set; }
		[Required]
		public string Company { get; set; }
		[Required]
		public int Phone { get; set; }
		[Required]
		public DateTime CreationDate { get; set; }
		[Required, ForeignKey("Device")]
		public int DeviceID { get; set; }
	}

	public enum WorkPlanType
	{
		Planned,
		Unplanned
	}

	public enum WorkPlanStatus
	{
		Draft,
		Approved,
		Denied,
		Canceled,
		Finished
	}
}
